### PR TITLE
ECL-465: Merge entity type and other entity type

### PR DESCRIPTION
**Pull request by paddynski-opencast**

### CODE REVIEW

File: app/uk/gov/hmrc/economiccrimelevyregistration/controllers/CheckYourAnswersController.scala [hunk 1]
```diff
@@ -107,29 +105,43 @@ class CheckYourAnswersController @Inject() (
           eclDetails(),
           organisationDetails(),
           contactDetails(),
+          otherEntityDetails(),
+          !request.registration.otherEntityJourneyData.isBlank(),
```
Comment (liandra-giacomine) [inline app/uk/gov/hmrc/economiccrimelevyregistration/controllers/CheckYourAnswersController.scala:109]: It looks like we only care about the case where it's defined, to avoid the negation you could change the function to `isDefined` and check for` _.isDefined `instead of `_.isEmpty`?

File: app/uk/gov/hmrc/economiccrimelevyregistration/controllers/EntityTypeController.scala [hunk 2]
```diff
@@ -70,17 +71,45 @@ class EntityTypeController @Inject() (
...
       )
   }
+
+  private def cleanup(
+    mode: Mode,
+    registration: Registration,
+    entityType: EntityType,
+    previousEntityType: Option[EntityType]
+  ) = {
+    val isOther = EntityType.isOther(entityType)
```
Comment (liandra-giacomine) [inline app/uk/gov/hmrc/economiccrimelevyregistration/controllers/EntityTypeController.scala:93]: This is a bit hard to follow, you could use pattern matching instead:

```
    (previousEntityType, mode) match {
      case (Some(entityType), CheckMode) if EntityType.isOther(entityType)          => registration
      case (Some(entityType), CheckMode)                                            =>
        dataCleanup.cleanup(
          registration.copy(
            entityType = Some(entityType)
          )
        )
      case (Some(entityType), _) if previousEntityType.contains(entityType) =>
        dataCleanup.cleanup(
          registration.copy(entityType = Some(entityType))
        )
      case _                                                                        =>
        dataCleanup.cleanupOtherEntityData(
          registration.copy(entityType = Some(entityType))
        )
    }

```

File: app/uk/gov/hmrc/economiccrimelevyregistration/controllers/CheckYourAnswersController.scala (hunk 1 above)
Comment (paddynski-opencast) [reply app/uk/gov/hmrc/economiccrimelevyregistration/controllers/CheckYourAnswersController.scala]: Done.

File: app/uk/gov/hmrc/economiccrimelevyregistration/controllers/EntityTypeController.scala (hunk 2 above)
Comment (liandra-giacomine) [reply app/uk/gov/hmrc/economiccrimelevyregistration/controllers/EntityTypeController.scala:93]: It looks like we're doing the same thing when previous entity type matches the entity type and when it doesn't so reading it it would seem that it would make most sense to have
 ```
case Some(Some(entityType),_) if EntityType.isOther(entityType) => 

          dataCleanup.cleanupOtherEntityData(
            registration.copy(entityType = Some(entityType))
          )
```

File: app/uk/gov/hmrc/economiccrimelevyregistration/controllers/CheckYourAnswersController.scala [hunk 3]
```diff
@@ -107,29 +105,43 @@ class CheckYourAnswersController @Inject() (
           eclDetails(),
           organisationDetails(),
           contactDetails(),
+          otherEntityDetails(),
+          request.registration.otherEntityJourneyData.isDefined(),
```
Comment (nemanjatomanovic) [inline app/uk/gov/hmrc/economiccrimelevyregistration/controllers/CheckYourAnswersController.scala]: We do not need this Boolean, because you can check previous parameter in the view. More details on view comment

File: app/uk/gov/hmrc/economiccrimelevyregistration/views/CheckYourAnswersView.scala.html [hunk 4]
```diff
@@ -26,6 +26,8 @@
 @(eclDetails: SummaryList,
         organisationDetails: SummaryList,
         contactDetails: SummaryList,
+        otherEntityDetails: SummaryList,
+        hasOtherEntityData: Boolean,
```
Comment (nemanjatomanovic) [inline app/uk/gov/hmrc/economiccrimelevyregistration/views/CheckYourAnswersView.scala.html:30]: We do not need this one, you can use otherEntityDetails summaryList and instead of boolean check do something like this
!otherEntityDetails.rows.isEmpty

File: app/uk/gov/hmrc/economiccrimelevyregistration/views/CheckYourAnswersView.scala.html [hunk 5]
```diff
@@ -50,6 +52,13 @@ <h1 class="govuk-heading-m">@messages("checkYourAnswers.contactDetails")</h1>
 
         @govukSummaryList(contactDetails)
 
+        @if(hasOtherEntityData) {
+
+            <h2>@messages("checkYourAnswers.otherEntityDetails")</h2>
```
Comment (nemanjatomanovic) [inline app/uk/gov/hmrc/economiccrimelevyregistration/views/CheckYourAnswersView.scala.html:57]: Previous comment applies

File: app/uk/gov/hmrc/economiccrimelevyregistration/views/CheckYourAnswersView.scala.html [hunk 6]
```diff
@@ -50,6 +52,13 @@ <h1 class="govuk-heading-m">@messages("checkYourAnswers.contactDetails")</h1>
 
         @govukSummaryList(contactDetails)
 
+        @if(hasOtherEntityData) {
```
Comment (nemanjatomanovic) [inline app/uk/gov/hmrc/economiccrimelevyregistration/views/CheckYourAnswersView.scala.html:55]: Previous comment applies.

File: app/uk/gov/hmrc/economiccrimelevyregistration/controllers/EntityTypeController.scala (hunk 2 above)
Comment (paddynski-opencast) [reply app/uk/gov/hmrc/economiccrimelevyregistration/controllers/EntityTypeController.scala:93]: I find pattern matching on several things at once more confusing.

File: app/uk/gov/hmrc/economiccrimelevyregistration/views/CheckYourAnswersView.scala.html (hunk 4 above)
Comment (liandra-giacomine) [reply app/uk/gov/hmrc/economiccrimelevyregistration/views/CheckYourAnswersView.scala.html:30]: Is there a `.isDefined` or `.nonEmpty` instead so we don't negate it?

Comment (nemanjatomanovic) [reply app/uk/gov/hmrc/economiccrimelevyregistration/views/CheckYourAnswersView.scala.html:30]: Oh yeah, for sure, it's a `Seq`,` .nonEmpty `would be perfect.

**Review by nemanjatomanovic — APPROVED** [review]
